measureme/src/profiling_data.rs

```diff
@@ -0,0 +1,73 @@
+use crate::event::Event;
+use crate::{ProfilerFiles, RawEvent, StringTable};
+use std::fs;
+use std::mem;
+use std::path::Path;
+use std::time::{Duration, SystemTime};
+
+pub struct ProfilingData {
+    event_data: Vec<u8>,
+    string_table: StringTable,
+}
+
+impl ProfilingData {
+    pub fn new(path_stem: &Path) -> ProfilingData {
+        let paths = ProfilerFiles::new(path_stem);
+
+        let string_data = fs::read(paths.string_data_file).expect("couldn't read string_data file");
+        let index_data = fs::read(paths.string_index_file).expect("couldn't read string_index file");
+        let event_data = fs::read(paths.events_file).expect("couldn't read events file");
+
+        let string_table = StringTable::new(string_data, index_data);
+
+        ProfilingData {
+            string_table,
+            event_data,
+        }
+    }
+
+    pub fn iter(&self) -> impl Iterator<Item = Event<'_>> {
+        ProfilerEventIterator {
+            data: self,
+            curr_event_idx: 0,
+        }
+    }
+}
+
+struct ProfilerEventIterator<'a> {
+    data: &'a ProfilingData,
+    curr_event_idx: usize,
+}
+
+impl<'a> Iterator for ProfilerEventIterator<'a> {
+    type Item = Event<'a>;
+
+    fn next(&mut self) -> Option<Event<'a>> {
+        let raw_idx = self.curr_event_idx * mem::size_of::<RawEvent>();
+        let raw_idx_end = raw_idx + mem::size_of::<RawEvent>();
+        if raw_idx_end > self.data.event_data.len() {
+            return None;
+        }
+
+        self.curr_event_idx += 1;
+
+        let raw_event_bytes = &self.data.event_data[raw_idx..raw_idx_end];
+
+        let raw_event = unsafe {
```

Comment:
Whoops, the old comment seems to be overwritten. See here: https://github.com/rust-lang/measureme/pull/5/commits/392d882aca536c4ce8502fdddbf25cb9de680355#r268630278

Reply:
Hmm... the link doesn't seem to work right for me so I'm not sure which comment you're referring to.

Reply:
Here's a copy:

We can't do this because the data at `raw_event_bytes` might not have the right alignment. In order to make this safe we have to allocate an event on the stack (e.g. `let mut event = RawEvent::default();`) and then copy the bytes from `raw_event_bytes` into the stack allocated `RawEvent`, i.e. the reverse of this:
https://github.com/rust-lang/measureme/blob/54c175dae6048433106d6099e446c6fd4bb898cd/src/profiler.rs#L59-L73

The important thing is that the `copy_from_slice` calls is done with `u8` slices (which are always safe to work with because of their alignment of 1) instead of pointers to `RawEvent`.